File: java/src/main/java/br/com/vps/consulting/b2b/management/partner/domain/Partner.java
package br.com.vps.consulting.b2b.management.partner.domain;

import br.com.vps.consulting.b2b.management.shared.core.entity.Entity;
import br.com.vps.consulting.b2b.management.shared.core.vo.Money;
import lombok.Builder;
import lombok.Getter;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Objects;

@Getter
public class Partner extends Entity<PartnerId> {

    private final PartnerId id;
    private final String name;
    private final String document;
    private final Instant createdAt;
    private Money creditLimit;

    @Builder
    public Partner(final PartnerId id,
                   final String name,
                   final String document,
                   final BigDecimal creditLimit,
                   final Instant createdAt) {
        this.id = id == null ? PartnerId.generate() : id;
        this.name = name;
        this.document = document;
        this.creditLimit = Money.of(creditLimit);
        this.createdAt = createdAt == null ? Instant.now() : createdAt;
        this.validate();
    }

    public void adjustCreditLimit(final BigDecimal newLimit) {
        Objects.requireNonNull(newLimit, "newLimit is required");
        this.creditLimit = Money.of(newLimit);
    }

    @Override
    protected void validate() {
        Objects.requireNonNull(this.id, "PartnerId is required");
        Objects.requireNonNull(this.name, "name is required");
        Objects.requireNonNull(this.document, "document is required");
        Objects.requireNonNull(this.creditLimit, "creditLimit is required");
        if (this.name.isBlank()) throw new IllegalArgumentException("name cannot be blank");
        if (this.document.isBlank()) throw new IllegalArgumentException("document cannot be blank");
    }

}

File: kotlin/src/main/kotlin/br/com/vps/consulting/b2b/management/partner/domain/Partner.kt
package br.com.vps.consulting.b2b.management.partner.domain

import br.com.vps.consulting.b2b.management.shared.core.entity.Entity
import java.time.Instant

class Partner private constructor(
    id: PartnerId,
    val name: String,
    val document: String,
    val createdAt: Instant,
) : Entity<PartnerId>(id) {

    init {
        validate()
    }

    override fun validate() {
        require(name.isNotBlank()) { "O nome não pode estar em branco" }
        require(document.isNotBlank()) { "O documento não pode estar em branco" }
    }

    companion object {
        fun create(name: String, document: String): Partner = Partner(
            id = PartnerId.generate(),
            name = name,
            document = document,
            createdAt = Instant.now(),
        )

        fun with(id: PartnerId, name: String, document: String, createdAt: Instant) =
            Partner(id, name, document, createdAt)
    }

}
